### DISCUSSION TITLE
High cpu usage / load balancer question

### DISCUSSION THREAD

**Discussion by hernanrz** [OvenMediaLabs/OvenMediaEngine#1193] [Help]
Hello, in our website with many input streams and hundreds of concurrent viewers, we see a very high cpu usage (nearing 90% on a 32 core cpu), we setup our website so LLHLS is the default source, the players (Ovenplayer) connect directly to OME, my question is: would adding a proxy, or load balancer like nginx in front of the OME server help us reduce the CPU usage?
[logs17apr.log](https://github.com/AirenSoft/OvenMediaEngine/files/11289199/logs17apr.log)
[Config.txt](https://github.com/AirenSoft/OvenMediaEngine/files/11289201/Config.txt)

This is a graph of cpu usage around the time we deployed the update, it goes back down once we made the players connect to nginx-rtmp HLS instead of OME llhls
![image](https://user-images.githubusercontent.com/13426225/233451867-ee99a735-f864-4d14-8cf1-cf9b6275398b.png)

**Comment by naanlizard** [comment]
For reference, many input streams is ~60 when we noticed this

**Comment by naanlizard** [comment]
@getroot I'm not sure if you get notifications for discussions, but this is something we're hoping for help with quite a lot

Thank you!

**Comment by getroot** [comment]
If there is a big difference in CPU usage between when there are many viewers accessing OvenMediaEngine's LLHLS and when there are not, it will of course be very helpful to use a CDN or HTTP cache server as an edge to distribute traffic. In this case, it is more accurate to describe it as an HTTP cache server rather than a load balancer.

**Reply by naanlizard** [reply]
Is it expected to have such high CPU usage with relatively few llhls clients? It seemed excessive to us. Any obvious settings to tweak?

**Reply by getroot** [reply]
What is the number of concurrent connections and output throughput?

**Reply by naanlizard** [reply]
I'd estimate 1000 clients max, probably less than that though.

By output throughput, do you mean bitrate? For each of 60 streams, average probably around 1.5mbps, no stream over 5mbps. I can find exact numbers if it will help

**Reply by getroot** [reply]
It would be helpful for me to analyze this issue if there is a measurement rather than an estimate of the network throughput.

**Reply by naanlizard** [reply]
<img width="717" alt="Screenshot 2023-04-26 at 7 50 30 AM" src="https://user-images.githubusercontent.com/7061721/234492778-e81f970c-d9b1-49dd-8adb-0eb44cf48dda.png">
<img width="710" alt="Screenshot 2023-04-26 at 7 50 42 AM" src="https://user-images.githubusercontent.com/7061721/234492830-620b2811-ed10-4460-af6e-9564e3250e5b.png">

<img width="697" alt="Screenshot 2023-04-26 at 7 51 17 AM" src="https://user-images.githubusercontent.com/7061721/234493221-a02e3b5b-8adb-453f-8973-be642e38e3c4.png">
<img width="711" alt="Screenshot 2023-04-26 at 7 51 35 AM" src="https://user-images.githubusercontent.com/7061721/234493436-0286643e-1d94-45c6-a838-f2e598754d44.png">

Here are several screenshots, the RTMP bandwidth is the sum total of incoming bandwidth (basically, bandwidth from nginx-rtmp to OME, as we forward each stream once this should be the total incoming bandwidth).

~117 mbps peak for rtmp input and ~~900mbps for output to viewers

The blue line in OME CPU usage is when we deployed the update and asked users to refresh

**Reply by getroot** [reply]
I'll discuss this with other team members and comment again. We do not have enough experience with LLHLS delivery because we mostly service it through a CDN.

**Reply by naanlizard** [reply]
Thank you, we'll potentially try caching through our nginx instance, but wanted to make sure we weren't misconfiguring something

**Reply by naanlizard** [reply]
@getroot  in the meantime can you share your Nginx cache configuration?

**Reply by getroot** [reply]
I dont use nginx for cache. I use CDN service like CF for LLHLS cache.

**Reply by naanlizard** [reply]
Ah of course, apologies. Can you share the company and pricing you get? We're looking for CDN options

**Reply by getroot** [reply]
In Korea, I use a local CDN company affiliated with us. And for global destinations, we use CloudFront, but there is no discount. :-(

**Reply by naanlizard** [reply]
Damn. Our problem is that we have a very good deal on bandwidth from our origin server +1, we would prefer to use that if possible. The performance with both our current nginx cache and straight to OME is not good, at least for llhls. We will continue to experiment, but any advice you can provide is very helpful

**Reply by getroot** [reply]
Is nginx not performing well when delivering LLHLS? Or is the cache hit rate low?

If your hit rate is low, consider the following.

I'm not familiar with nginx, but most edge sets the cache duration of a file with the HTTP Cache-Control header that origin responds to.

OvenMediaEnigne sends HTTP Cache-Control header in origin mode, which can be set as follows.

0 means no-cache.
-1 means don't send Cache-Control header. In this case, the CDN caches files for as many as its settings.
0 < means max-age:

```
<LLHLS>
	<OriginMode>true</OriginMode>
	<CacheControl>
		<MasterPlaylistMaxAge>0</MasterPlaylistMaxAge>
		<ChunklistMaxAge>0</ChunklistMaxAge>
		<ChunklistWithDirectivesMaxAge>60</ChunklistWithDirectivesMaxAge>
		<SegmentMaxAge>-1</SegmentMaxAge>
		<PartialSegmentMaxAge>-1</PartialSegmentMaxAge>
	</CacheControl>
```

And one of the main roles of Edge is to send only one of the multiple requests for the same file to the origin and make the other requests wait. I'm not sure if nginx does this well.

**Reply by hernanrz** [reply]
In a attempt to increase our cache hit rate in nginx (right now I'm confident all of the requests result in a miss), we changed the cache key setting to the following:

```
            proxy_cache_key $schema$proxy_host$uri;
```

This is so it would ignore query string parameters in the URL, for caching, as a result we were getting a lot of cache hits from nginx but now playback wasn't working for llhls, any tips on what we did wrong here would be appreciated

**Reply by getroot** [reply]
@hernanrz 1. Did you set OriginMode to true in OME?
2. I don't know about nginx setup. However, chunklists without directives in LLHLS should not be cached. That way, past chunklists can be passed to the player.


URLs without directives such as `https://host/app/stream/llhls.m3u8` should always deliver the latest chunklist. So it cannot be cached.

URLs with directives such as `https://host/app/stream/llhls.m3u8?_HLS_msn=10&_HLS_part=1902` can be cached.

**Reply by getroot** [reply]
The best way is to determine cache availability and duration from the cache-control header delivered by OME LLHLS. But your nginx configuration doesn't seem to do that.

**Comment by naanlizard** [comment]
I will bring this thread back to life a little - you've said you do many large livestreams @getroot, how do you handle caching for them? 

One downside of using nginx as a proxy/cache for OME is that when we export stats from the API via [oven-exporter](https://github.com/genofire/oven-exporter) to our grafana instance, we don't get accurate client counts (as clients appear to all connect from the proxy) 
<img width="704" alt="Screenshot 2023-10-31 at 2 10 50 PM" src="https://github.com/AirenSoft/OvenMediaEngine/assets/7061721/f413d551-77c7-4ecf-836f-66bb003f892c">

Is there a way to have accurate stats from OME, while also maintaining good performance? Perhaps OME should eventually implement some sort of caching layer itself?

**Reply by getroot** [reply]
When using a cache server such as CDN, information on the number of concurrent users can only be provided by the cache server. (Because only the first session comes into Origin).

Of course, that is not very easy unless Cache provides such functionality. The same goes for CloudFront. So, most of us find the approximate number of concurrent users through network throughput. (If 2Gbps throughput occurs when serving at 2Mbps, it is calculated as about 1,000 people.) When our customers request the exact number of concurrent users, there were cases where we had the player transmit metrics for statistics.

**Reply by naanlizard** [reply]
Alright, I figured as much but was hoping there was something I was missing. Our bandwidth usage is mixed with other site stuff, and of course we have many input streams, so estimating bandwidth per stream is tougher.

**Reply by getroot** [reply]
We also have a variety of services mixed in our CDN, but we divide them by domain and extract statistics such as throughput.

**Comment by basisbit** [comment]
Just my 2ct to this topic: From my experience, you can forget about the usual CDN-caching approach for when using LLHLS or WebRTC. 

Instead, either use HLS (which seems to be somewhat broken in current releases of OME - workaround: have HLS on port 80, and LLHLS on port 443, then point your CDN to port 80 of the OME server) to be able to use the usual approach of forced-caching of the `*.ts` files for a minute on the CDN.

Or, if you want low latency, don't use a HTTP cache CDN in front of your OME server, but instead build your own CDN: place a few OME edge servers in each region/continent of your users, and implement a load distributor yourself (for example forwarding the HTTP GET request for the playlist based on IP address of the HTTP request origin). You can use Maxmind lookup database to find the best suitable region to deliver video from. Also track how many users are pulling streams from which edge server and load-balance as needed. Adding more servers where needed can also be automated with this. I have implemented this approach for a small side-project during Corona in my free time, and it worked very well for events with more than 50k stream viewers, at low cost. (much lower than using one of the big CDNs)

Also, see https://github.com/AirenSoft/OvenMediaEngine/issues/343

**Reply by ciroiriarte** [reply]
always wondered how to catch content locally. So, instead of going with a CDN, the edge nodes can act as CDN/cache?

**Reply by ciroiriarte** [reply]
on that line of thought, will the edge nodes synchronize content within a region?.

scenario would be, X edge servers in region A and Y in region B because of the viewers demand, I don't want X+Y requests for the same content going all the way to the origin.

**Reply by basisbit** [reply]
A CDN is a collection if servers which deliver the content to the users. There is no need to cache anything.
The way how the Origin-Edge-structure works is: someone sends the video stream to an origin server, which transcodes the video (if enabled/configured) and then forwards it to the webrtc/ovt/srt/rtmp clients or provides it for the HLS/LLHLS providers. Edge servers will usually receive the stream over ovt from the origin server.

Your website/webserver has to handle giving your viewers the best fitting WebRTC/HLS/LLHLS address to grab the stream from.
Having some servers in the different regions of the world and having clients get the video from local region is a common thing to do. There is no "synchronize content" necessary, because all the edge servers pull the streams from the origin servers on demand if you use OVT.
